### DISCUSSION TITLE
Closing server in Docker : gracefull shutdown

### DISCUSSION THREAD

**Discussion by averatio** [aio-libs/aiohttp#7571] [Q&A] [answered]
Hello Community
I run an aiohttp server in a Docker container orchestrated by Kubernetes. I need a gracefull shutdown on Docker Term signal.
The application object documentation says: "It’s up to end user to figure out which [web-handler](https://docs.aiohttp.org/en/v3.8.1/glossary.html#term-web-handler)s are still alive and how to finish them properly."
Does aiohttp framework provide a way to list the running handlers ?
Shall I write a specific code to register instances in a set within the app object to track executions?
Advices are welcome
AV

**Comment by Dreamsorcerer** [comment]
I'm not entirely clear what you're trying to achieve. That documentation is talking about things like websockets, which are persistent connections. If you use websockets, then you'd typically store all the websockets in a set, so that during shutdown, you can ask them all to gracefully close. There are plenty of examples around, such as the chat demo: https://github.com/aio-libs/aiohttp-demos/tree/master/demos/chat/aiohttpdemo_chat

Note in 3.9, we've also added a new feature that will wait a few seconds for running tasks before cancelling them. So, for regular web handlers that last no more than a few seconds, they should have the time to complete normally once using aiohttp 3.9 (which will hopefully have a beta release in the next few weeks).

**Reply by averatio** [reply]
Thank you @Dreamsorcerer. I have no websocket but handlers which transfer heavy files for longer that tens of seconds I want to prevent the Docker Term sign to close the connection while transfering. Additionally, I use aiojibs with background tasks that I don't want to interrupt on Term signal. I would like to shutdown the server when all handlers and tasks are terminated. May be the framework does it but it's unclear to me.

**Reply by Dreamsorcerer** [reply]
Then probably you want to keep a reference to those tasks and await on them in the shutdown handler (or you can change the wait timeout in aiohttp 3.9+).

**Reply by averatio** [reply]
Yes, exactly, but I wonder whether loop.run_until_complete(loop.shutdown_asyncgens()) is sufficient as suggested [here](https://stackoverflow.com/questions/71062375/aiohttp-terminate-web-app-run). I would like to avoid tasks individual execution tracking.

**Reply by Dreamsorcerer** [reply, marked as answer]
I don't think that's a suggestion, that's a copy of a bit of our source code. You shouldn't use the low-level asyncio APIs for general use, and we'd likely have already cancelled the tasks before you could run that anyway.

The only way to reliably achieve it, is to track the tasks you want to keep running (similar to the the websocket example) and await them all in the shutdown handler (also, you probably want to make sure it's the first shutdown handler, to avoid DB etc. being closed before the handlers complete).

**Reply by Dreamsorcerer** [reply]
Or you could just use a really long `shutdown_timeout`: https://docs.aiohttp.org/en/latest/web_reference.html#aiohttp.web.run_app
And hope that you don't end up with any misbehaving tasks that block your shutdown...

**Reply by averatio** [reply]
Thank you,  I will write a handler tracker so that I await handlers end before closing the server.
